File: sdks/ruby/lib/novabilling/types/cohort_row.rb
# frozen_string_literal: true

module Novabilling
  module Types
    class CohortRow < Internal::Types::Model
      field :month, -> { String }, optional: false, nullable: false
      field :total_customers, -> { Integer }, optional: false, nullable: false, api_name: "totalCustomers"
      field :retention_percentages, -> { Internal::Types::Array[Integer] }, optional: false, nullable: false, api_name: "retentionPercentages"
    end
  end
end

File: sdks/csharp/src/NovaBilling/Types/CohortRow.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record CohortRow : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    [JsonPropertyName("month")]
    public required string Month { get; set; }

    [JsonPropertyName("totalCustomers")]
    public required double TotalCustomers { get; set; }

    [JsonPropertyName("retentionPercentages")]
    public IEnumerable<double> RetentionPercentages { get; set; } = new List<double>();

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}
